# Lab 1
## Step 1 Naive Bayes
### Import packages and data

In [25]:
import os
import pandas as pd
import numpy as np
from pandas.api.types import CategoricalDtype
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB
from sklearn.feature_extraction.text import CountVectorizer

PATH_ROOT = os.getcwd()
PATH_TRAIN = os.path.join(PATH_ROOT, 'train.csv')
PATH_TEST = os.path.join(PATH_ROOT, 'test.csv')

print("Use train data:", PATH_TRAIN)
print("Use test  data:", PATH_TRAIN)

train_data = pd.read_csv(PATH_TRAIN)
test_data = pd.read_csv(PATH_TEST)

Use train data: d:\Windows\Documents\Code\conda\machine-learning-lab\lab1\train.csv
Use test  data: d:\Windows\Documents\Code\conda\machine-learning-lab\lab1\train.csv


### Recoginize categories

In [26]:
categories = train_data['category'].unique()
print(categories)

['Restaurants' 'Nightlife' 'Shopping']


We find that there are three categories in the data set, they are: `['Restaurants' 'Nightlife' 'Shopping']`.

Then, we transform this column into integers.

In [27]:
categories_type = CategoricalDtype(categories = categories)
train_data['category'] = train_data['category'].astype(categories_type).cat.codes.astype('long')
test_data['category'] = test_data['category'].astype(categories_type).cat.codes.astype('long')
print(train_data['category'].head(), "\n", test_data['category'].head(), sep="")

0    0
1    0
2    0
3    1
4    0
Name: category, dtype: int32
0    0
1    0
2    0
3    0
4    0
Name: category, dtype: int32


### Naive Bayes process
We start our Naive Bayes process.

Firstly, we should build training and testing dataframe variables.


In [28]:
train_y = train_data['category']
train_x = train_data['review']

test_y = test_data['category']
test_x = test_data['review']

Then we need to build a vector of word counts. Use built in class `CountVectorizer`. And transform original data into vector.

In [29]:
vector = CountVectorizer()
train_x = vector.fit_transform(train_x).toarray()
test_x = vector.transform(test_x).toarray()

Other attributes also need to be considered. We merge these data into training and testing data set.

In [30]:
train_x = np.append(train_data[['latitude', 'longitude', 'mean_checkin_time']], train_x, axis=1)
test_x = np.append(test_data[['latitude', 'longitude', 'mean_checkin_time']], test_x, axis=1)

We have already built our training and testing data set.

Lastly, we could classify texts by using Naive Bayes.

In [31]:
nb = GaussianNB()
nb.fit(train_x, train_y)

GaussianNB()

We can see how many correct prediction we have made.

In [32]:
nb.score(test_x,test_y)

0.7532467532467533

About $75.3%$ of the entire data set has been classified correctly.

## Step 2 Optimization